curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/OutstandingOps.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.curator.framework.recipes.cache;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+class OutstandingOps
```

Comment:
An alternative is that `java.util.concurrent.Phaser` which already implements semantic like `OutstandingOps`.

```java
Runnable r = () -> callListeners(CuratorCacheListener::initialized);
Phaser p = new Phaser() {
  @Override
            protected boolean onAdvance(int phase, int registeredParties) {
                if (registeredParties > 0) {
                    return false;
                }
                r.run();
                return true;
            }
}
```

and replace `increment` as `register` while `decrement` as `arriveAndDeregister`

Reply:
Ooh - TIL - I didn't know about that. Thanks. Looking into it now.

Reply:
Phaser looks to be a kind of barrier. How would it be used in this context @TisonKun ?

Reply:
It seems `OutstandingOps` is a one-shot barrier, so that as the code snippet above when we have new outstanding op we register a party and terminate the phaser at the first time it is advanced. Is it the semantic of `OutstandingOps`?

Reply:
It's the same idea as in TreeCache. When the cache is started each outstanding operation increments a counter and when the op completes it decrements the counter. It's not exact, but it provides a way to handle the "cache is initialized" feature. Once the outstanding ops counter gets back to 0 it's no longer used.

Reply:
Yes that is the same as things in my mind. Since `OutstandingOps` implemented correctly now, we can have a follow-up to see if it's worth we replace it with Java builtins and I will share a patch with description there. At most it is an "improvement" and should not block this PR to merge :)

For the "how", it seems to me that https://github.com/apache/curator/pull/335#discussion_r399591042 is clear.